### PR TITLE
Refactor apiserver endpoint transformers to more natively use Encoders

### LINKED ISSUE #83898: Refactor transformObject to implement Encoder interface
Long-term, transformObject:
https://github.com/kubernetes/kubernetes/blob/5219ad7be209b31cdfdae2eda1cd435aff4a3cd2/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L40
should be refactored so that it implements Encoder interface (and it has nested encoder inside):

This would allow us to cache the results for difference media.Convert options, which currently is not supported:
https://github.com/kubernetes/kubernetes/blob/5219ad7be209b31cdfdae2eda1cd435aff4a3cd2/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L50

@kubernetes/sig-api-machinery-misc @liggitt

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix https://github.com/kubernetes/kubernetes/issues/83898

This PR on its own addresses the above issue.
But its main purpose is to enable a large memory allocation optimization:
- in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1152-less-object-serializations we created a framework and avoided serializing the same object multiple times
- however, there is one big piece that I missed:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L274-L290
- even though the serialization of watch-event is simple [we just copy bytes around], we're allocating quite a bit of memory there and optimizing that can bring non-negligible savings
- this PR [in addition to being a nice cleanup on its own] enables that improvement in a clear way

```release-note
NONE
```

To reviewers: I recommend reviewing commit-by-commit - commits are self-contained and build on top of each other.

/kind cleanup
/priority important-longterm
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/helper.go [hunk 1]
```diff
@@ -257,3 +257,26 @@ func (d WithoutVersionDecoder) Decode(data []byte, defaults *schema.GroupVersion
 	}
 	return obj, gvk, err
 }
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/runtime/helper.go:260]: I recommend reviewing commit-by-commit - commits are self-contained and build on top of each other.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 2]
```diff
@@ -29,30 +32,116 @@ import (
...
+	identifier runtime.Identifier
+}
+
+func newWatchEmbeddedEncoder(ctx context.Context, encoder runtime.Encoder, target *schema.GroupVersionKind, options interface{}, scope *RequestScope) *watchEmbeddedEncoder {
+	result := &watchEmbeddedEncoder{
+		encoder: encoder,
+		ctx:     ctx,
+		target:  target,
+	}
+	if target != nil && target.Kind == "Table" {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:67]: *metav1.TableOptions is the only options type we expect here, so require a positive check and fail if this is non-nil and not *metav1.TableOptions (or change the signature here to take `options *metav1.TableOptions` and check at the caller)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 3]
```diff
@@ -29,30 +32,116 @@ import (
...
+	result, err := doTransformObject(e.ctx, obj, e.options, e.target, e.scope)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to transform object %v: %v", reflect.TypeOf(obj), err))
+		result = obj
+	}
+
+	// When we are tranforming to a table, use the original table options when
+	// we should print headers only on the first object - headers should be
+	// omitted on subsequent events.
+	if tableOptions, ok := e.options.(*metav1.TableOptions); ok && !tableOptions.NoHeaders {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:94]: switch check to non-nil check once the options field type changes

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 4]
```diff
@@ -29,30 +32,116 @@ import (
...
+	Target  string              `json:"target,omitempty"`
+	Options metav1.TableOptions `json:"options,omitempty"`
+	Scope   string              `json:"scope,omitempty"`
+}
+
+func (e *watchEmbeddedEncoder) embeddedIdentifier() runtime.Identifier {
+	if e.target == nil {
+		// If no conversion is performed, we effective only use
+		// the embedded identifier.
+		return e.encoder.Identifier()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:124]: add a comment to the target == nil case in `doTransformObject` that if we ever change that from a no-op, this needs to update to include whatever variable behavior we have

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 5]
```diff
@@ -29,30 +32,116 @@ import (
...
+	}
+	identifier := watchEmbeddedEncoderIdentifier{
+		Name:    "watch-embedded",
+		Encoder: string(e.encoder.Identifier()),
+		Target:  e.target.String(),
+	}
+	if e.target.Kind == "Table" {
+		if options, ok := e.options.(*metav1.TableOptions); ok && options != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:132]: switch this to a non-nil check

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 6]
```diff
@@ -29,30 +32,116 @@ import (
...
+		e.identifier = e.embeddedIdentifier()
+	}
+	return e.identifier
+}
+
+type watchEmbeddedEncoderIdentifier struct {
+	Name    string              `json:"name,omitempty"`
+	Encoder string              `json:"encoder,omitempty"`
+	Target  string              `json:"target,omitempty"`
+	Options metav1.TableOptions `json:"options,omitempty"`
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:116]: `NoHeaders bool `json:"-"`` means this won't serialize distinguishing headers from non-headers

that also means tests aren't exercising the header/non-header caching behavior

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 7]
```diff
@@ -29,30 +32,116 @@ import (
...
+		Encoder: string(e.encoder.Identifier()),
+		Target:  e.target.String(),
+	}
+	if e.target.Kind == "Table" {
+		if options, ok := e.options.(*metav1.TableOptions); ok && options != nil {
+			identifier.Options = *options
 		}
+		// TODO(wojtek-t): Can we improve it?
+		//   Is this really needed - can it differ across requests?
+		identifier.Scope = fmt.Sprintf("%p;%p", e.scope.TableConvertor, e.scope.Convertor)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:137]: drop this from the identifier

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 8]
```diff
@@ -60,15 +60,6 @@ func doTransformObject(ctx context.Context, obj runtime.Object, opts interface{}
 		return obj, nil
 	}
 
-	// ensure that for empty lists we don't return <nil> items.
-	// This is safe to modify without deep-copying the object, as
-	// List objects themselves are never cached.
-	if meta.IsListType(obj) && meta.LenList(obj) == 0 {
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:66]: Done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 9]
```diff
@@ -106,6 +105,26 @@ func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions n
 		embeddedEncoder = scope.Serializer.EncoderForVersion(serializer.Serializer, contentKind.GroupVersion())
 	}
 
+	var memoryAllocator runtime.MemoryAllocator
+
+	if encoderWithAllocator, supportsAllocator := embeddedEncoder.(runtime.EncoderWithAllocator); supportsAllocator {
+		// don't put the allocator inside the embeddedEncodeFn as that would allocate memory on every call.
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:117]: I tested that against some integration tests with the following diff:
```
diff --git a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go b/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go
index ab874e30bfe..45bf5be20b4 100644
--- a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go
+++ b/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go
@@ -34,6 +34,8 @@ import (
        "k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
        "k8s.io/apiserver/pkg/endpoints/metrics"
        apirequest "k8s.io/apiserver/pkg/endpoints/request"
+
+       "k8s.io/klog/v2"
 )
 
 // nothing will ever be sent down this channel
@@ -113,6 +115,7 @@ func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions n
                // instead, we allocate the buffer for the entire watch session and release it when we close the connection.
                memoryAllocator = runtime.AllocatorPool.Get().(*runtime.Allocator)
                defer runtime.AllocatorPool.Put(memoryAllocator)
+               klog.Errorf("MEMORY_ALOCATOR: embedded encoder")
                embeddedEncoder = runtime.NewEncoderWithAllocator(encoderWithAllocator, memoryAllocator)
        }
        if tableOptions, ok := options.(*metav1.TableOptions); ok {
@@ -128,6 +131,7 @@ func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions n
                        memoryAllocator = runtime.AllocatorPool.Get().(*runtime.Allocator)
                        defer runtime.AllocatorPool.Put(memoryAllocator)
                }
+               klog.Errorf("MEMORY_ALOCATOR: encoder")
                encoder = runtime.NewEncoderWithAllocator(encoderWithAllocator, memoryAllocator)
        }
```

It works as expected.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:119]: done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go (hunk 6 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go]: Thanks for offline suggestion about the test Jordan - it made the trick to achieve it [I extended the test that I added before to handle that].

**Comment by wojtek-t** [comment]
@liggitt - comments addressed, PTAL

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 10]
```diff
@@ -114,6 +115,11 @@ func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions n
 		defer runtime.AllocatorPool.Put(memoryAllocator)
 		embeddedEncoder = runtime.NewEncoderWithAllocator(encoderWithAllocator, memoryAllocator)
 	}
+	var tableOptions *metav1.TableOptions
+	if passedOptions, ok := options.(*metav1.TableOptions); ok {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:119]: if all code paths that propagate options here give us TableOptions, should this treat a non-nil non-TableOptions as an error? if we ever add another options type, I want to make sure we don't forget this piece

**Pull request merged**